frame/vesting/src/lib.rs

```diff
@@ -61,41 +60,136 @@ use frame_support::traits::{
+		starting_block: BlockNumber,
+	}
+
+	impl<Balance: AtLeast32BitUnsigned + Copy, BlockNumber: AtLeast32BitUnsigned + Copy>
+		VestingInfo<Balance, BlockNumber>
+	{
+		/// Instantiate a new `VestingInfo`.
+		pub fn new<T: Config>(
+			locked: Balance,
+			per_block: Balance,
+			starting_block: BlockNumber,
+		) -> VestingInfo<Balance, BlockNumber> {
+			VestingInfo { locked, per_block, starting_block }
+		}
+
+		/// Validate parameters for `VestingInfo`. Note that this does not check
+		/// against `MinVestedTransfer` or the current block. Additionally it
+		/// will correct `per_block` if it is greater than `locked`.
+		pub fn validate<T: Config>(mut self) -> Result<Self, Error<T>> {
+			ensure!(
+				!self.locked.is_zero() && !self.per_block.is_zero(),
+				Error::<T>::InvalidScheduleParams
+			);
+			self.per_block =
+				if self.per_block > self.locked { self.locked } else { self.per_block };
+			Ok(self)
+		}
+
+		/// Instantiate a new `VestingInfo` without param modification. Useful for
+		/// mocking bad inputs in testing.
+		#[cfg(test)]
+		pub fn unsafe_new(
+			locked: Balance,
+			per_block: Balance,
+			starting_block: BlockNumber,
+		) -> VestingInfo<Balance, BlockNumber> {
+			VestingInfo { locked, per_block, starting_block }
+		}
+
+		/// Locked amount at schedule creation.
+		pub fn locked(&self) -> Balance {
+			self.locked
+		}
+
+		/// Amount that gets unlocked every block after `starting_block`.
+		pub fn per_block(&self) -> Balance {
+			self.per_block
+		}
+
+		/// Starting block for unlocking(vesting).
+		pub fn starting_block(&self) -> BlockNumber {
+			self.starting_block
+		}
+
+		/// Amount locked at block `n`.
+		pub fn locked_at<BlockNumberToBalance: Convert<BlockNumber, Balance>>(
+			&self,
+			n: BlockNumber,
+		) -> Balance {
+			// Number of blocks that count toward vesting
+			// Saturating to 0 when n < starting_block
+			let vested_block_count = n.saturating_sub(self.starting_block);
+			let vested_block_count = BlockNumberToBalance::convert(vested_block_count);
+			// Return amount that is still locked in vesting
+			vested_block_count.checked_mul(&self.per_block)
+                .map(|balance| self.locked.saturating_sub(balance))
+                .unwrap_or(Zero::zero())
+		}
+
+		/// Block number at which the schedule ends.
+		pub fn ending_block<BlockNumberToBalance: Convert<BlockNumber, Balance>>(&self) -> Balance {
+			let starting_block = BlockNumberToBalance::convert(self.starting_block);
+			let duration = if self.per_block > self.locked {
+				// If `per_block` is bigger than `locked`, the schedule will end
+				// the block after starting.
+				1u32.into()
+			} else if self.per_block.is_zero() {
+				// Check for div by 0 errors, which should only be from legacy
+				// vesting schedules since new ones are validated for this.
+				self.locked
+			} else {
+				let has_remainder = !(self.locked % self.per_block).is_zero();
+				let maybe_duration = self.locked / self.per_block;
```

Comment:
```suggestion
				let mut maybe_duration = self.locked / self.per_block;
```

I don't know how this is compiling now :D